clang/lib/Basic/DiagnosticIDs.cpp

```diff
@@ -575,25 +575,42 @@ DiagnosticIDs::getDiagnosticSeverity(unsigned DiagID, SourceLocation Loc,
       DiagID != diag::fatal_too_many_errors && Diag.FatalsAsError)
     Result = diag::Severity::Error;
 
+  // Rest of the mappings are only applicable for diagnostics associated with a
+  // SourceLocation, bail out early for others.
+  if (!Diag.hasSourceManager())
```

Comment:
NIT: maybe split into a separate review and land right away?

This looks pretty orthogonal to the rest of the patch, so it'd make for a better commit history.